engine/src/main/java/org/terasology/rendering/nui/editor/NUIEditorContextMenuBuilder.java

```diff
@@ -0,0 +1,350 @@
+ */
+package org.terasology.rendering.nui.editor;
+
+import com.google.api.client.util.Lists;
+import com.google.api.client.util.Maps;
+import com.google.gson.annotations.SerializedName;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.terasology.persistence.ModuleContext;
+import org.terasology.reflection.metadata.ClassMetadata;
+import org.terasology.reflection.metadata.FieldMetadata;
+import org.terasology.rendering.nui.AbstractWidget;
+import org.terasology.rendering.nui.LayoutConfig;
+import org.terasology.rendering.nui.NUIManager;
+import org.terasology.rendering.nui.UILayout;
+import org.terasology.rendering.nui.UIWidget;
+import org.terasology.rendering.nui.contextMenu.ContextMenuBuilder;
+import org.terasology.rendering.nui.contextMenu.ContextMenuLevel;
+import org.terasology.rendering.nui.databinding.Binding;
+import org.terasology.rendering.nui.skin.UISkin;
+import org.terasology.rendering.nui.widgets.UpdateListener;
+import org.terasology.rendering.nui.widgets.treeView.JsonTree;
+import org.terasology.rendering.nui.widgets.treeView.JsonTreeValue;
+import org.terasology.utilities.ReflectionUtil;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.ParameterizedType;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+
+public class NUIEditorContextMenuBuilder {
+
+    private Logger logger = LoggerFactory.getLogger(NUIEditorContextMenuBuilder.class);
+
+    // Context menu levels.
+    public static final String LEVEL_ADD_EXTENDED = "Add";
+    public static final String LEVEL_PRIMARY = "Primary";
+
+    // Context menu options.
+    public static final String OPTION_ADD_EXTENDED = "Add...";
+    public static final String OPTION_ADD_WIDGET = "Add Widget";
+    public static final String OPTION_COPY = "Copy";
+    public static final String OPTION_EDIT = "Edit";
+    public static final String OPTION_PASTE = "Paste";
+
+    /**
+     * A {@link NUIManager} instance retrieved from the editor screen.
+     */
+    private NUIManager nuiManager;
+    /**
+     * External consumer methods to be used when creating the menu.
+     */
+    private Map<String, Consumer<JsonTree>> externalConsumers = Maps.newHashMap();
+    /**
+     * Listeners fired when one of the options within {@code LEVEL_ADD_EXTENDED} is selected.
+     */
+    private List<UpdateListener> addContextMenuListeners = Lists.newArrayList();
+
+    public void setManager(NUIManager nuiManager) {
+        this.nuiManager = nuiManager;
+    }
+
+    public void putConsumer(String key, Consumer<JsonTree> value) {
+        externalConsumers.put(key, value);
+    }
+
+    public void subscribeAddContextMenu(UpdateListener listener) {
+        addContextMenuListeners.add(listener);
+    }
+
+    public ContextMenuBuilder createPrimaryContextMenu(JsonTree node) {
+        ContextMenuBuilder contextMenu = new ContextMenuBuilder();
+        ContextMenuLevel primaryLevel = contextMenu.addLevel(LEVEL_PRIMARY);
```

Comment:
The constants are temporary either way since I'm planning to apply i18n to the editor after all the strings are finalized. Using constants instead of inline values would make the refactor easier.